lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -282,12 +287,33 @@ instance LiftHandler ErrCreateUnsignedTx where
 
 instance LiftHandler ErrSignTx where
     handler = \case
-        ErrSignTx _ -> err500
+        ErrSignTx (KeyNotFoundForAddress addr) -> err403
+            { errBody =
+                "Error signing transaction: corresponding private key for the \
+                \following output address was not found: " <> pretty addr
+            }
         ErrSignTxNoSuchWallet _ -> err410
         ErrSignTxWithRootKey e -> handler e
 
 instance LiftHandler ErrSubmitTx where
-    handler _ = err500
+    handler = \case
+        ErrSubmitTxNetwork e -> case e of
+            ErrPostTxNetworkUnreachable (ErrNetworkUnreachable err) -> err500
+                { errBody =
+                    "Err submitting transaction: network is unreachable: "
+                    <> pretty err
+                }
+            ErrPostTxBadRequest err -> err500
+                { errBody =
+                    "Err submitting transaction: request rejected by node: "
+                    <> pretty err
+                }
+            ErrPostTxProtocolFailure err -> err500
+                { errBody =
+                    "Err submitting transaction: protocol failure: "
+                    <> pretty err
+                }
+        ErrSubmitTxNoSuchWallet _ -> err404
```

Comment:
NOTE: Ideally, we should do this for pretty much all errors as well as making sure we add more context to our errors!

Reply:
This is good to have.

For an API consumer it is nicer to have error messages as JSON responses (e.g. `{ "status": "error", "message": "Submiting transaction: protocol failure" }`). Then the API consumer can always parse all responses as JSON, rather than worrying about whether it will be `text/plain` or `application/json`.

Reply:
Indeed. Ideally, something more like `{ "code": <an_error_code>, "message": <description / message> }`, the status is redundant with the HTTP status. However, having a clear discriminant is helpful for clients like Daedalus (instead of having to grep on the message).